### PR TITLE
twister: fix timeout status for the device handler

### PR DESCRIPTION
**Pull request by evgeniy-paltsev**
Currently in binary handler and qemu handler we have status `failed` in case of test timeout, but in device handler we have status `error`. This not only adds inconsistency between handlers, but also prevents us from usage test retry functionality for the runs on HW.

Fix timeout status by changing it to `failed` instead of `error`.

### CODE REVIEW

**Review by abrodkin — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Comment by PerMac** [comment]
I believe such change is against our convention. Failures are reserved for actual failed asserts. Other issues are errors. Such inconsistency should be resolved the other way. I will have a quick look to check the retry issue.

**Review by PerMac — CHANGES_REQUESTED** [review]
I believe such change is against our convention. Failures are reserved for actual failed asserts. Other issues are errors. Such inconsistency should be resolved the other way. I will have a quick look to check the retry issue.

**Comment by PerMac** [comment]
It seems we might need to clean up statuses, I see more inconsistencies. E.g.
https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib/handlers.py#L536
and https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib/handlers.py#L575
I believe the idea behind not retrying errors is due to the fact that we don't want to retry build errors by default. There was a flag added `--retry-build-errors` but the implementation makes all errors to retry, since there is no dedicated status for build error. https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib/runner.py#L1069

To sum up, I think we need to define and wrote down in docs definitions for all possible statuses. Also, introduce a dedicated build_error status. Then, evaluate all places in the code where statuses are assigned, to unify their usage. 
@nashif Opinions?

**Review by nashif — APPROVED** [review]
I think this particular case should be indeed be a FAIL and we need to change the option name `--retry-build-errors` to be `--retry-errors`.

**Comment by PerMac** [comment]
What about https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib/handlers.py#L575
``` 
    if proc.returncode != 0:
        self.instance.status = "error"
        self.instance.reason = "Device issue (Flash error?)"
```
I assume this is a good candidate to be retried by default. Would the solution be to also switch it to "failed"? 
Build errors are IMO specific enough to get own status. In vast majority such errors are deterministic, and there is no point in retrying (I know about builds that downloads some libs, those can be prone to random connection issues). Retrying all build errors along other errors sounds like sth that can significantly slow down on-target CIs, where more "retriable errors" can appear.

**Comment by nashif** [comment]
> I assume this is a good candidate to be retried by default. Would the solution be to also switch it to "failed"?

no, this is still an error. 
Lets not mix too many things. The status being changed here is not different from other timeouts in other handlers.


> Build errors are IMO specific enough to get own status. In vast majority such errors are deterministic, and there is no point in retrying (I know about builds that downloads some libs, those can be prone to random connection issues).

Build errors are VERY common if your toolchain has to connect to a license server and fails, so retrying is needed. I think keeping the status as is keeps things consistent. The retry options needs to be modified and fixed to allow exact selection of what you want to rety based on your environment and we can address this in another PR. I agree that retrying all build failures by default is not wanted and needs to be optional an separated from other errors.

**Review by PerMac — APPROVED** [review]
